addons/digest/tests/test_digest.py

```diff
@@ -0,0 +1,75 @@
+from unittest.mock import patch
+import random
+
+from dateutil.relativedelta import relativedelta
+from lxml import html
+
+from odoo.addons.mail.models.mail_mail import MailMail
+
+from odoo import fields
+from odoo.tests import TransactionCase
+
+
+class TestDigestStuff(TransactionCase):
+    def test_digest_numbers(self):
+        self._setup_messages()
+
+        d = self.env['digest.digest'].create({'name': "My Digest", 'kpi_mail_message_total': True})
+
+        # subscribe the demo user so at least one mail gets sent
+        demo = self.env.ref('base.user_demo')
+        d_demo = d.with_env(self.env(user=demo))
+        d_demo.action_subscribe()
+        self.assertTrue(
+            d_demo.is_subscribed,
+            "check the user was subscribed as action_subscribe will silently "
+            "ignore subs of non-employees"
+        )
+
+        # digest creates its mails in auto_delete mode so we need to capture
+        # the formatted body during the sending process
+        body = None
+        def send(self, *args, _original=MailMail.send, **kwargs):
+            nonlocal body
+            body = self.body_html
+            return _original(self, *args, **kwargs)
+
+        with patch.object(MailMail, 'send', autospec=True, wraps=MailMail, side_effect=send):
+            d.action_send()
+        self.assertTrue(body, "a mail has been created for the digest")
+
+        kpi_values = html.fromstring(body).xpath('//*[hasclass("kpi_cell")]//*[hasclass("kpi_value")]')
+        self.assertEqual(
+            [v.text.strip() for v in kpi_values],
+            ['3', '8', '15']
+        )
+
+    def _setup_messages(self):
+        """ Remove all existing messages, then create a bunch of them on random
+        partners with the correct types in correct time-bucket:
+
+        - 3 in the previous 24h
+        - 5 more in the 6 days before that for a total of 8 in the previous week
+        - 7 more in the 20 days before *that* (because digest doc lies and is
+          based around weeks and months not days), for a total of 15 in the
+          previous month
+        """
+        self.env['mail.message'].search([]).unlink()
+        now = fields.Datetime.now()
+        partners = self.env['res.partner'].search([])
```

Comment:
Meh. Actually why do you need partners at all ?

Reply:
To post messages on them?

Reply:
Arf. Can't you post on some newly creted mail.channel instead ? Looks like "more real use case" oriented.

Reply:
Does that actually work? The message KPI looks for subtype "mt_comment" and message type "comment" or "email".